File: src/main/java/net/easecation/ghosty/recording/player/updated/PlayerUpdatedPing.java
package net.easecation.ghosty.recording.player.updated;

import cn.nukkit.utils.BinaryStream;
import cn.nukkit.utils.TextFormat;
import net.easecation.ghosty.entity.PlaybackNPC;
import net.easecation.ghosty.recording.player.PlayerRecordNode;

/**
 * Created by Mulan Lin('Snake1999') on 2016/11/19 17:02.
 * All rights reserved
 */
public class PlayerUpdatedPing implements PlayerUpdated {

    private int ping;

    public static PlayerUpdatedPing of(int ping) {
        return new PlayerUpdatedPing(ping);
    }

    @Override
    public int getUpdateTypeId() {
        return PlayerUpdated.TYPE_PING;
    }

    @Override
    public boolean hasStates() {
        return false;
    }

    @Override
    public void processTo(PlaybackNPC ghost) {
        ghost.lastPing = this.ping;
        if (ghost.getEngine().displayPlayerPing) {
            String[] split = ghost.getNameTag().split("\n");
            ghost.setNameTag(getDisplayPing(this.ping) + TextFormat.WHITE + "\n" + split[split.length - 1]);
        }
    }

    public static String getDisplayPing(int ping) {
        if (ping < 0) {
            return "N/A";
        } else if (ping < 200) {
            return TextFormat.GREEN.toString() + ping + "ms";
        } else if (ping < 500) {
            return TextFormat.YELLOW.toString() + ping + "ms";
        } else {
            return TextFormat.RED.toString() + ping + "ms";
        }
    }

    @Override
    public PlayerRecordNode applyTo(PlayerRecordNode node) {
        node.setPing(this.ping);
        return node;
    }

    public PlayerUpdatedPing(BinaryStream stream) {
        this.ping = (int) stream.getUnsignedVarInt();
    }

    private PlayerUpdatedPing(int ping) {
        this.ping = ping;
    }

    @Override
    public boolean equals(Object obj) {
        if(!(obj instanceof PlayerUpdatedPing o)) return false;
        return ping == o.ping;
    }

    @Override
    public void write(BinaryStream stream) {
        stream.putUnsignedVarInt(this.ping);
    }

    @Override
    public void read(BinaryStream stream) {
        this.ping = (int) stream.getUnsignedVarInt();
    }

    @Override
    public String toString() {
        return "PlayerUpdatedPing{" +
                "ping=" + ping +
                '}';
    }
}

File: src/main/java/net/easecation/ghosty/recording/player/updated/PlayerUpdatedPing.kt
package net.easecation.ghosty.recording.player.updated

import cn.nukkit.utils.BinaryStream
import cn.nukkit.utils.TextFormat.*
import kotlinx.serialization.Serializable
import net.easecation.ghosty.entity.PlaybackNPC
import net.easecation.ghosty.recording.player.PlayerRecordNode

@Serializable
data class PlayerUpdatedPing(
    val ping: Int
) : PlayerUpdated {

    override fun getUpdateTypeId() = PlayerUpdated.TYPE_PING

    override fun processTo(ghost: PlaybackNPC) {
        ghost.lastPing = ping
        if (ghost.engine?.displayPlayerPing == true) {
            val parts = ghost.nameTag.split("\n")
            ghost.nameTag = "${getDisplayPing(ping)}${WHITE}\n${parts.lastOrNull()}"
        }
    }

    override fun applyTo(node: PlayerRecordNode) = node.also {
        it.ping = this.ping
    }

    companion object {
        @JvmStatic
        fun of(ping: Int) = PlayerUpdatedPing(ping)

        @JvmStatic
        fun getDisplayPing(ping: Int): String = when {
            ping < 0 -> "N/A"
            ping < 200 -> "${GREEN}${ping}ms"
            ping < 500 -> "${YELLOW}${ping}ms"
            else -> "${RED}${ping}ms"
        }

        @JvmStatic
        val ADAPTER: PlayerUpdateAdapter<PlayerUpdatedPing> = Adapter
    }

    private object Adapter : PlayerUpdateAdapter<PlayerUpdatedPing> {
        override fun write(updated: PlayerUpdatedPing, stream: BinaryStream) {
            stream.putUnsignedVarInt(updated.ping.toLong())
        }

        override fun read(stream: BinaryStream): PlayerUpdatedPing {
            return PlayerUpdatedPing(stream.unsignedVarInt.toInt())
        }
    }
}
